src/java.base/share/classes/jdk/internal/foreign/abi/s390/linux/LinuxS390CallArranger.java

```diff
@@ -0,0 +1,304 @@
+            new VMStorage[] { r2, r3, r4, r5, r6, }, // GP input
+            new VMStorage[] { f0, f2, f4, f6 }, // FP input
+            new VMStorage[] { r2, }, // GP output
+            new VMStorage[] { f0, }, // FP output
+            new VMStorage[] { r0, r1, r2, r3, r4, r5, r14 }, // volatile GP
+            new VMStorage[] { f1, f3, f5, f7 }, // volatile FP (excluding argument registers)
+            8, // Stack is always 8 byte aligned on S390
+            160, // ABI header
+            r0, r1 // scratch reg r0 & r1
+            );
+
+    public record Bindings(CallingSequence callingSequence, boolean isInMemoryReturn) {}
+
+    public static Bindings getBindings(MethodType mt, FunctionDescriptor cDesc, boolean forUpcall) {
+        return getBindings(mt, cDesc, forUpcall, LinkerOptions.empty());
+    }
+
+    public static Bindings getBindings(MethodType mt, FunctionDescriptor cDesc, boolean forUpcall, LinkerOptions options) {
+        CallingSequenceBuilder csb = new CallingSequenceBuilder(CLinux, forUpcall, options);
+
+        BindingCalculator argCalc = forUpcall ? new BoxBindingCalculator(true) : new UnboxBindingCalculator(true);
+        BindingCalculator retCalc = forUpcall ? new UnboxBindingCalculator(false) : new BoxBindingCalculator(false);
+
+        boolean returnInMemory = isInMemoryReturn(cDesc.returnLayout());
+        if (returnInMemory) {
+            Class<?> carrier = MemorySegment.class;
+            MemoryLayout layout =SharedUtils.C_POINTER;
+            csb.addArgumentBindings(carrier, layout, argCalc.getBindings(carrier, layout));
+        } else if (cDesc.returnLayout().isPresent()) {
+            Class<?> carrier = mt.returnType();
+            MemoryLayout layout = cDesc.returnLayout().get();
+            csb.setReturnBindings(carrier, layout, retCalc.getBindings(carrier, layout));
+        }
+
+        for (int i = 0; i < mt.parameterCount(); i++) {
+            Class<?> carrier = mt.parameterType(i);
+            MemoryLayout layout = cDesc.argumentLayouts().get(i);
+            csb.addArgumentBindings(carrier, layout, argCalc.getBindings(carrier, layout));
+        }
+
+        return new Bindings(csb.build(), returnInMemory);
+    }
+
+    /* Replace DowncallLinker in OpenJDK with the implementation of DowncallLinker specific to OpenJ9 */
+    public static MethodHandle arrangeDowncall(MethodType mt, FunctionDescriptor cDesc, LinkerOptions options) {
+        return DowncallLinker.getBoundMethodHandle(mt, cDesc, options);
+    }
+
+    public static UpcallStubFactory arrangeUpcall(MethodType mt, FunctionDescriptor cDesc, LinkerOptions options) {
+        Bindings bindings = getBindings(mt, cDesc, true, options);
+
+        final boolean dropReturn = true; /* drop return, since we don't have bindings for it */
+        return SharedUtils.arrangeUpcallHelper(mt, bindings.isInMemoryReturn, dropReturn, CLinux,
+                bindings.callingSequence);
+    }
```

Comment:
I think this should be taken from `s390x/sysv/CallArranger.java`:
```java
	/* Replace UpcallLinker in OpenJDK with the implementation of UpcallLinker specific to OpenJ9. */
	public static UpcallStubFactory arrangeUpcall(MethodType mt, FunctionDescriptor cDesc, LinkerOptions options) {
		return UpcallLinker.makeFactory(mt, cDesc, options);
	}
```

Reply:
Fixed `arrangeUpcall` as well.

Also removed `SysVPPC64le` and `SysVS390x` from `test/jdk/java/foreign/callarranger/platform/PlatformLayouts.java`.